File: CS/AgendaView/AgendaViewDataGenerator.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using DevExpress.XtraScheduler;
using System.ComponentModel;
using System.Drawing;

namespace AgendViewComponent {
    public static class AgendaViewDataGenerator {
        public static TimeInterval SelectedInterval { get; set; }

        public static object GenerateResourcesCollection(SchedulerStorage storage) {
            return storage.Resources.Items;        
        }

        public static object GenerateAgendaAppointmentCollection(SchedulerStorage storage) {
            AppointmentBaseCollection sourceAppointments = storage.GetAppointments(SelectedInterval);
            BindingList<AgendaAppointment> agendaAppointments = new BindingList<AgendaAppointment>();
            foreach(Appointment appointment in sourceAppointments) {
                TimeInterval currentDayInterval = new TimeInterval(appointment.Start.Date, appointment.Start.Date.AddDays(1));
                string startTime = "";
                string endTime = "";
                if(currentDayInterval.Contains(appointment.End)) {
                    startTime = currentDayInterval.Start == appointment.Start ? "" : appointment.Start.TimeOfDay.ToString(@"hh\:mm");
                    endTime = currentDayInterval.End == appointment.End ? "" : appointment.End.TimeOfDay.ToString(@"hh\:mm");
                    agendaAppointments.Add(CreateAgendaAppointment(storage, appointment, currentDayInterval.Start, startTime, endTime));
                }
                else {
                    startTime = currentDayInterval.Start == appointment.Start ? "" : appointment.Start.TimeOfDay.ToString(@"hh\:mm");
                    agendaAppointments.Add(CreateAgendaAppointment(storage, appointment, currentDayInterval.Start, startTime, ""));
                    while(true) {
                        currentDayInterval = new TimeInterval(currentDayInterval.End, currentDayInterval.End.AddDays(1));
                        if(currentDayInterval.Contains(appointment.End)) {
                            endTime = currentDayInterval.End == appointment.End ? "" : appointment.End.TimeOfDay.ToString(@"hh\:mm");
                            agendaAppointments.Add(CreateAgendaAppointment(storage, appointment, currentDayInterval.Start, "", endTime));
                            break;
                        }
                        else {
                            agendaAppointments.Add(CreateAgendaAppointment(storage, appointment, currentDayInterval.Start, "", ""));
                        }
                    }

                }
            }
            return agendaAppointments;
        }

        static AgendaAppointment CreateAgendaAppointment(SchedulerStorage storage, Appointment sourceAppointment, DateTime startDate, string startTime, string endTime) {
            AgendaAppointment agendaAppointment = new AgendaAppointment();
            agendaAppointment.AgendaDate = startDate;
            agendaAppointment.AgendaDescription = sourceAppointment.Description;
            agendaAppointment.AgendaSubject = sourceAppointment.Subject;
            if(startTime == "" && endTime == "") {
                agendaAppointment.AgendaDuration = "All Day";
            }
            else if(startTime == "" && endTime != "") {
                agendaAppointment.AgendaDuration = "Till: " + endTime;
            }
            else if(startTime != "" && endTime == "") {
                agendaAppointment.AgendaDuration = "From: " + startTime;
            }
            else {
                agendaAppointment.AgendaDuration = String.Format("{0} - {1}", startTime, endTime);
            }

            agendaAppointment.AgendaLocation = sourceAppointment.Location;
            agendaAppointment.AgendaStatus = storage.Appointments.Statuses[sourceAppointment.StatusId]; ;
            agendaAppointment.AgendaLabel = storage.GetLabelColor(sourceAppointment.LabelId);
            agendaAppointment.SourceAppointment = sourceAppointment;
            return agendaAppointment;
        }
    }

    // Agenda appointment
    public class AgendaAppointment {
        public AppointmentStatus AgendaStatus { get; set; }
        public string AgendaSubject { get; set; }
        public string AgendaDescription { get; set; }
        public string AgendaDuration { get; set; }
        public string AgendaLocation { get; set; }
        public DateTime AgendaDate { get; set; }
        public Color AgendaLabel { get; set; }
        public Appointment SourceAppointment { get; set; }
    }
}

File: VB/AgendaView/AgendaViewDataGenerator.vb
Imports System
Imports System.Collections.Generic
Imports System.Linq
Imports System.Text
Imports DevExpress.XtraScheduler
Imports System.ComponentModel
Imports System.Drawing

Namespace AgendViewComponent
    Public NotInheritable Class AgendaViewDataGenerator
        Public Shared Property SelectedInterval() As TimeInterval

        Private Sub New()
        End Sub


        Public Shared Function GenerateResourcesCollection(ByVal storage As SchedulerStorage) As Object
            Return storage.Resources.Items
        End Function

        Public Shared Function GenerateAgendaAppointmentCollection(ByVal storage As SchedulerStorage) As Object
            Dim sourceAppointments As AppointmentBaseCollection = storage.GetAppointments(SelectedInterval)
            Dim agendaAppointments As New BindingList(Of AgendaAppointment)()
            For Each appointment As Appointment In sourceAppointments
                Dim currentDayInterval As New TimeInterval(appointment.Start.Date, appointment.Start.Date.AddDays(1))
                Dim startTime As String = ""
                Dim endTime As String = ""
                If currentDayInterval.Contains(appointment.End) Then
                    startTime = If(currentDayInterval.Start = appointment.Start, "", appointment.Start.TimeOfDay.ToString("hh\:mm"))
                    endTime = If(currentDayInterval.End = appointment.End, "", appointment.End.TimeOfDay.ToString("hh\:mm"))
                    agendaAppointments.Add(CreateAgendaAppointment(storage, appointment, currentDayInterval.Start, startTime, endTime))
                Else
                    startTime = If(currentDayInterval.Start = appointment.Start, "", appointment.Start.TimeOfDay.ToString("hh\:mm"))
                    agendaAppointments.Add(CreateAgendaAppointment(storage, appointment, currentDayInterval.Start, startTime, ""))
                    Do
                        currentDayInterval = New TimeInterval(currentDayInterval.End, currentDayInterval.End.AddDays(1))
                        If currentDayInterval.Contains(appointment.End) Then
                            endTime = If(currentDayInterval.End = appointment.End, "", appointment.End.TimeOfDay.ToString("hh\:mm"))
                            agendaAppointments.Add(CreateAgendaAppointment(storage, appointment, currentDayInterval.Start, "", endTime))
                            Exit Do
                        Else
                            agendaAppointments.Add(CreateAgendaAppointment(storage, appointment, currentDayInterval.Start, "", ""))
                        End If
                    Loop

                End If
            Next appointment
            Return agendaAppointments
        End Function

        Private Shared Function CreateAgendaAppointment(ByVal storage As SchedulerStorage, ByVal sourceAppointment As Appointment, ByVal startDate As Date, ByVal startTime As String, ByVal endTime As String) As AgendaAppointment
            Dim agendaAppointment As New AgendaAppointment()
            agendaAppointment.AgendaDate = startDate
            agendaAppointment.AgendaDescription = sourceAppointment.Description
            agendaAppointment.AgendaSubject = sourceAppointment.Subject
            If startTime = "" AndAlso endTime = "" Then
                agendaAppointment.AgendaDuration = "All Day"
            ElseIf startTime = "" AndAlso endTime <> "" Then
                agendaAppointment.AgendaDuration = "Till: " & endTime
            ElseIf startTime <> "" AndAlso endTime = "" Then
                agendaAppointment.AgendaDuration = "From: " & startTime
            Else
                agendaAppointment.AgendaDuration = String.Format("{0} - {1}", startTime, endTime)
            End If

            agendaAppointment.AgendaLocation = sourceAppointment.Location
            agendaAppointment.AgendaStatus = storage.Appointments.Statuses(sourceAppointment.StatusId)

            agendaAppointment.AgendaLabel = storage.GetLabelColor(sourceAppointment.LabelId)
            agendaAppointment.SourceAppointment = sourceAppointment
            Return agendaAppointment
        End Function
    End Class

    ' Agenda appointment
    Public Class AgendaAppointment
        Public Property AgendaStatus() As AppointmentStatus
        Public Property AgendaSubject() As String
        Public Property AgendaDescription() As String
        Public Property AgendaDuration() As String
        Public Property AgendaLocation() As String
        Public Property AgendaDate() As Date
        Public Property AgendaLabel() As Color
        Public Property SourceAppointment() As Appointment
    End Class
End Namespace
